File: src/problems/week-3/t_primes.cpp
#include <iostream>
#include <math.h>

using namespace std;

int is_prime(int n) {
  if (n <= 3) {
    return n > 1;
  }
  if (n % 2 == 0) {
    return 0;
  }

  for (long int i = 3; i < (int) sqrt(n) + 1; i += 2) {
    if (n % i == 0) {
      return 0;
    }
  }
  return 1;
}

int is_t_prime(long long int n) {
  long long int root = sqrt(n);
  if (root * root != n) {
    return 0;
  }

  if (is_prime(root)) {
    return 1;
  }
  return 0;
}

int main() {
  long int n;
  cin >> n;

  while (n--) {
    long long int number;
    cin >> number;
    if (is_t_prime(number)) {
      cout << "YES\n";
    } else {
      cout << "NO\n";
    }
  }

  return 0;
}

File: src/problems/week-3/t_primes.py
from math import sqrt

n = int(input())
numbers = list(map(int, input().split()))

def is_prime(n: int):
  if n <= 3:
    return n > 1
  if n % 2 == 0:
    return False
  
  for i in range(3, int(sqrt(n)) + 1, 2):
    if n % i == 0:
      return False
  return True

# def creat_prime(n: int):
#   primes = []
#   for i in range(0, n):
#     pass

def is_t_primes(n: int):
  root = int(sqrt(n))
  if root ** 2 != n:
    return False
  
  if is_prime(root):
    return True
  return False

for number in numbers:
  if is_t_primes(number):
    print("YES")
  else:
    print("NO")
